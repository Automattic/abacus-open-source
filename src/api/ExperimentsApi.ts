import _ from 'lodash'
import * as yup from 'yup'

import {
  ExperimentFull,
  ExperimentFullNew,
  experimentFullNewOutboundSchema,
  experimentFullNewSchema,
  experimentFullSchema,
  ExperimentSummary,
  experimentSummaryResponse,
  MetricAssignmentNew,
  metricAssignmentNewOutboundSchema,
  metricAssignmentNewSchema,
  Status,
  yupPick,
} from 'src/lib/schemas'
import { isDebugMode } from 'src/utils/general'

import { fetchApi } from './utils'

/**
 * Attempts to create a new experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newExperiment: ExperimentFullNew): Promise<ExperimentFull> {
  const validatedNewExperiment = await experimentFullNewSchema.validate(newExperiment, { abortEarly: false })
  const outboundNewExperiment = experimentFullNewOutboundSchema.cast(validatedNewExperiment)
  return await experimentFullSchema.validate(await fetchApi('POST', '/experiments', outboundNewExperiment))
}

/**
 * Attempts to PUT an experiment, overwriting the existing experiment.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(experimentId: number, experiment: ExperimentFullNew): Promise<ExperimentFull> {
  const validatedExperiment = await experimentFullNewSchema.validate(experiment, { abortEarly: false })
  const outboundExperiment = experimentFullNewOutboundSchema.cast(validatedExperiment)
  return await experimentFullSchema.validate(await fetchApi('PUT', `/experiments/${experimentId}`, outboundExperiment))
}

/**
 * Attempts to patch an experiment.
 *
 * Doesn't work with nested fields.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function patch(experimentId: number, experimentPatch: Partial<ExperimentFull>): Promise<ExperimentFull> {
  // We dynamically construct a schema for validation, but we do this simply and shallowly
  const dynamicValidationSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch))
  const validatedExperimentPatch = await dynamicValidationSchema.validate(experimentPatch, { abortEarly: false })

  // Similarly we dynamically construct a schema for outbound casting
  const dynamicOutboundCastSchema = yupPick(experimentFullSchema, Object.keys(experimentPatch)).snakeCase()
  const outboundExperimentPatch = dynamicOutboundCastSchema.cast(validatedExperimentPatch)

  return await experimentFullSchema.validate(
    await fetchApi('PATCH', `/experiments/${experimentId}`, outboundExperimentPatch),
  )
}

async function changeStatus(experimentId: number, status: Status): Promise<void> {
  await fetchApi('PUT', `/experiments/${experimentId}/status`, { status })
}

/**
 * Attempts to assign a metric to an experiment.
 *
 * Metric Assignment is a bit funky right now, so we do it here.
 * In particular it require passing in all existing metricAssignments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @param experiment A full experiment object, needed to fill in the existing metricAssignments, will be removed in the future.
 * @param metricAssignment The new metricAssignment.
 */
async function assignMetric(
  experiment: ExperimentFull,
  metricAssignment: MetricAssignmentNew,
): Promise<ExperimentFull> {
  const validatedMetricAssignment = await metricAssignmentNewSchema.validate(metricAssignment)
  const outboundMetricAssignment = metricAssignmentNewOutboundSchema.cast(validatedMetricAssignment)

  // The backend replaces all the metricAssignments, so we format it as metricAssignmentNewOutbound.
  const outboundExistingMetricAssignments = yup
    .array(metricAssignmentNewOutboundSchema)
    .cast(experiment.metricAssignments.map((metricAssignment) => _.omit(metricAssignment, 'metricAssignmentId')))
  // istanbul ignore next; Shouldn't occur
  if (!outboundExistingMetricAssignments) {
    throw new Error('Something went wrong while transforming existing metricAssignments to outbound form.')
  }

  const metricAssignments = [...outboundExistingMetricAssignments, outboundMetricAssignment]

  return await experimentFullSchema.validate(
    await fetchApi('PATCH', `/experiments/${experiment.experimentId}`, {
      metric_assignments: metricAssignments,
    }),
  )
}

/**
 * Finds all the available experiments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<ExperimentSummary[]> {
  // istanbul ignore next; debug only
  const { experiments } = await experimentSummaryResponse.validate(
    await fetchApi('GET', isDebugMode() ? '/experiments?debug=true' : '/experiments'),
    { abortEarly: false },
  )
  return experiments
}

/**
 * Fetches the experiment with full details.
 *
 * @param id - The ID of the experiment to fetch.
 */
async function findById(id: number): Promise<ExperimentFull> {
  return await experimentFullSchema.validate(await fetchApi('GET', `/experiments/${id}`), { abortEarly: false })
}

const ExperimentsApi = {
  create,
  put,
  patch,
  assignMetric,
  changeStatus,
  findAll,
  findById,
}

export default ExperimentsApi
