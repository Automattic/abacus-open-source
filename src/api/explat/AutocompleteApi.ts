import { StatusCodes } from 'http-status-codes'

import { fetchApi } from 'src/api/explat/utils'
import { AutocompleteItem, autocompleteSchema, eventDetailsSchema } from 'src/lib/explat/schemas'

import HttpResponseError from '../HttpResponseError'

async function getCompletion(name: string) {
  return await autocompleteSchema.validate(await fetchApi('GET', `/autocomplete/${name}`), { abortEarly: false })
}

export async function getUserCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('users')).completions
}

export async function getEventNameCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('events')).completions
}

export async function getPropNameCompletions(eventName: string): Promise<AutocompleteItem[] | null> {
  if (!eventName) {
    throw new Error('No eventName to getPropNameCompletions of.')
  }

  try {
    const apiResponse = await eventDetailsSchema.validate(await fetchApi('GET', `/autocomplete/events/${eventName}`))
    return apiResponse.props.map((p) => ({
      name: p.name,
      value: p.name,
    }))
  } catch (error) {
    // istanbul ignore else; Forced to use the else here to prevent two istanbul ignores
    if (error instanceof HttpResponseError && error.status === StatusCodes.NOT_FOUND) {
      return null
    } else {
      throw error
    }
  }
}
