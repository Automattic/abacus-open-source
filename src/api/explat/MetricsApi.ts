import _ from 'lodash'
import * as yup from 'yup'

import {
  Metric,
  MetricNew,
  MetricNewOutbound,
  metricNewOutboundSchema,
  metricNewSchema,
  metricSchema,
} from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Transforms a metric into an outbound form.
 */

function outboundTransformMetric(metricNew: MetricNew): MetricNewOutbound {
  const metricDataForRequest = {
    ...metricNew,
    tags: metricNew.tags.map((tagId) => ({ tagId })),
  }
  return metricNewOutboundSchema.cast(metricDataForRequest)
}

/**
 * Attempts to create a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newMetric: MetricNew): Promise<Metric> {
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('POST', '/metrics', outboundNewMetric))
}

/**
 * Attempts to put a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(metricId: number, newMetric: MetricNew): Promise<Metric> {
  // istanbul ignore next; Shouldn't happen
  if (!_.isNumber(metricId)) {
    throw new Error('Invalid metricId.')
  }
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('PUT', `/metrics/${metricId}`, outboundNewMetric))
}

/**
 * Finds all the available metrics.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(options: { includeDebug: boolean } = { includeDebug: false }): Promise<Metric[]> {
  // istanbul ignore next; debug only
  const { metrics } = await yup
    .object({ metrics: yup.array(metricSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', options.includeDebug ? '/metrics?debug=true' : '/metrics'), {
      abortEarly: false,
    })
  return metrics
}

/**
 * Find the metric by ID.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findById(metricId: number): Promise<Metric> {
  return await metricSchema.validate(await fetchApi('GET', `/metrics/${metricId}`), { abortEarly: false })
}

const MetricsApi = {
  create,
  put,
  findAll,
  findById,
}

export default MetricsApi
