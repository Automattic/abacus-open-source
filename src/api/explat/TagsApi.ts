import _ from 'lodash'
import * as yup from 'yup'

import {
  TagBare,
  tagBareSchema,
  TagFull,
  TagFullNew,
  tagFullNewOutboundSchema,
  tagFullNewSchema,
  tagFullSchema,
} from 'src/lib/explat/schemas'
import { isDebugMode } from 'src/utils/general'

import { fetchApi } from './utils'

/**
 * Attempts to create a new tag.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newTag: TagFullNew): Promise<TagBare> {
  const validatedNewTag = await tagFullNewSchema.validate(newTag, { abortEarly: false })
  const outboundNewTag = tagFullNewOutboundSchema.cast(validatedNewTag)
  return await tagBareSchema.validate(await fetchApi('POST', '/tags', outboundNewTag))
}

/**
 * Attempts to put a new tag.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(tagId: number, newTag: TagFullNew): Promise<TagBare> {
  // istanbul ignore next; Shouldn't happen
  if (!_.isNumber(tagId)) {
    throw new Error('Invalid tagId.')
  }
  const validatedNewTag = await tagFullNewSchema.validate(newTag, { abortEarly: false })
  const outboundNewTag = tagFullNewOutboundSchema.cast(validatedNewTag)
  return await tagBareSchema.validate(await fetchApi('PUT', `/tags/${tagId}`, outboundNewTag))
}

/**
 * Finds all the available tags.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<TagBare[]> {
  // istanbul ignore next; debug only
  const { tags } = await yup
    .object({ tags: yup.array(tagBareSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', isDebugMode() ? '/tags?debug=true' : '/tags'), {
      abortEarly: false,
    })
  return tags
}

/**
 * Find the tag by ID.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findById(tagId: number): Promise<TagFull> {
  return await tagFullSchema.validate(await fetchApi('GET', `/tags/${tagId}`), { abortEarly: false })
}

const TagsApi = {
  create,
  put,
  findAll,
  findById,
}

export default TagsApi
