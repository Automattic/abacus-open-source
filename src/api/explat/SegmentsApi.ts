import * as yup from 'yup'

import { Segment, segmentSchema } from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Finds all the available segments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<Segment[]> {
  const { segments } = await yup
    .object({ segments: yup.array(segmentSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', '/segments'), { abortEarly: false })
  return segments
}

const SegmentsApi = {
  findAll,
}

export default SegmentsApi
