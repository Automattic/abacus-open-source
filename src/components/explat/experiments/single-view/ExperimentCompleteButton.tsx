import {
  Button,
  createStyles,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  makeStyles,
  Tooltip,
  Typography,
} from '@material-ui/core'
import { useSnackbar } from 'notistack'
import React, { useState } from 'react'

import ExperimentsApi from 'src/api/explat/ExperimentsApi'
import { serverErrorMessage } from 'src/api/HttpResponseError'
import PrivateLink from 'src/components/general/PrivateLink'
import { ExperimentFull, Status } from 'src/lib/explat/schemas'
import { useDangerStyles } from 'src/styles/styles'

import LoadingButtonContainer from '../../../general/LoadingButtonContainer'

const useStyles = makeStyles(() =>
  createStyles({
    dangerImage: {
      textAlign: 'center',
    },
  }),
)

const ExperimentCompleteButton = ({
  experiment,
  experimentReloadRef,
}: {
  experiment: ExperimentFull | null
  experimentReloadRef: React.MutableRefObject<() => void>
}): JSX.Element => {
  const classes = useStyles()
  const dangerClasses = useDangerStyles()
  const { enqueueSnackbar } = useSnackbar()

  const canCompleteExperiment = experiment && experiment.status === Status.Running
  const [isAskingToConfirmCompleteExperiment, setIsAskingToConfirmCompleteExperiment] = useState<boolean>(false)
  const onAskToConfirmCompleteExperiment = () => setIsAskingToConfirmCompleteExperiment(true)
  const onCancelCompleteExperiment = () => setIsAskingToConfirmCompleteExperiment(false)
  const [isSubmittingCompleteExperiment, setIsSubmittingCompleteExperiment] = useState<boolean>(false)
  const onConfirmCompleteExperiment = async () => {
    try {
      // istanbul ignore next; Shouldn't occur
      if (!experiment) {
        throw Error('Missing experiment, this should not happen')
      }

      setIsSubmittingCompleteExperiment(true)
      await ExperimentsApi.changeStatus(experiment.experimentId, Status.Completed)
      enqueueSnackbar('Experiment Completed!', { variant: 'success' })
      experimentReloadRef.current()
      setIsAskingToConfirmCompleteExperiment(false)
    } catch (e) /* istanbul ignore next; Shouldn't occur */ {
      console.log(e)
      enqueueSnackbar(`Oops! Something went wrong while trying to complete your experiment. ${serverErrorMessage(e)}`, {
        variant: 'error',
      })
    } finally {
      setIsSubmittingCompleteExperiment(false)
    }
  }

  return (
    <>
      <Tooltip
        title={
          canCompleteExperiment
            ? 'Stop new user assignments. Current assignments remain active.'
            : `This experiment is ${experiment?.status ?? 'undefined status'}.`
        }
      >
        <span>
          <Button
            variant='outlined'
            classes={{ outlined: dangerClasses.dangerButtonOutlined }}
            disabled={!canCompleteExperiment}
            onClick={onAskToConfirmCompleteExperiment}
          >
            2. Complete
          </Button>
        </span>
      </Tooltip>
      <Dialog
        open={isAskingToConfirmCompleteExperiment}
        aria-labelledby='confirm-complete-experiment-dialog-title'
        BackdropProps={{ className: dangerClasses.dangerBackdrop }}
      >
        <DialogTitle>
          <Typography variant='h5' component='div'>
            Are you sure you want to <strong>complete</strong> this experiment?
          </Typography>
        </DialogTitle>
        <DialogContent>
          <Typography variant='body2' gutterBottom>
            Completing will{' '}
            <strong>
              prevent new users from being assigned to the experiment, users that have already been assigned will
              continue with their assigned experience.
            </strong>{' '}
            This may take up to ten minutes to propagate to all servers due to{' '}
            <PrivateLink
              href='https://wp.me/PCYsg-Fq9#logged-out-homepage-assignments-use-file-system-cache'
              rel='noopener noreferrer'
              target='_blank'
            >
              the file system assignment cache
            </PrivateLink>
            .
          </Typography>
          <Typography variant='body2' gutterBottom>
            Completing also changes the experiment&apos;s status to completed, which is <strong>irreversible</strong>.
          </Typography>
          <div className={classes.dangerImage}>
            <img src='/img/danger.gif' alt='DANGER!' />
          </div>
        </DialogContent>
        <DialogActions>
          <Button variant='contained' color='primary' onClick={onCancelCompleteExperiment}>
            Cancel
          </Button>
          <LoadingButtonContainer isLoading={isSubmittingCompleteExperiment}>
            <Button
              variant='contained'
              classes={{ contained: dangerClasses.dangerButtonContained }}
              disabled={isSubmittingCompleteExperiment}
              onClick={onConfirmCompleteExperiment}
            >
              Complete
            </Button>
          </LoadingButtonContainer>
        </DialogActions>
      </Dialog>
    </>
  )
}

export default ExperimentCompleteButton
