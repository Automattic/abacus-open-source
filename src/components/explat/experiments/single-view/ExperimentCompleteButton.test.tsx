/* eslint-disable @typescript-eslint/require-await */
import { fireEvent, screen, waitFor, waitForElementToBeRemoved } from '@testing-library/react'
import MockDate from 'mockdate'
import React from 'react'

import ExperimentsApi from 'src/api/explat/ExperimentsApi'
import { Status } from 'src/lib/explat/schemas'
import Fixtures from 'src/test-helpers/fixtures'
import { render } from 'src/test-helpers/test-utils'

import ExperimentCompleteButton from './ExperimentCompleteButton'

MockDate.set('2020-07-21')

jest.mock('src/api/explat/ExperimentsApi')
const mockedExperimentsApi = ExperimentsApi as jest.Mocked<typeof ExperimentsApi>

test('renders as expected', () => {
  const experimentReloadRef: React.MutableRefObject<() => void> = { current: jest.fn() }
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentCompleteButton {...{ experiment, experimentReloadRef }} />)

  expect(container).toMatchSnapshot()
})

test('completes an experiment', async () => {
  const experimentReloadRef: React.MutableRefObject<() => void> = { current: jest.fn() }
  const experiment = Fixtures.createExperimentFull({ status: Status.Running })
  const { container } = render(<ExperimentCompleteButton {...{ experiment, experimentReloadRef }} />)

  mockedExperimentsApi.changeStatus.mockReset()
  mockedExperimentsApi.changeStatus.mockImplementationOnce(async () => undefined)

  const firstCompleteButton = screen.getByRole('button', { name: /Complete/ })

  // First Opening - We cancel
  fireEvent.click(firstCompleteButton)

  await waitFor(() => screen.getByRole('button', { name: /Cancel/ }))

  expect(container).toMatchSnapshot()

  const cancelButton = screen.getByRole('button', { name: /Cancel/ })
  fireEvent.click(cancelButton)
  await waitForElementToBeRemoved(cancelButton)

  expect(mockedExperimentsApi.changeStatus).toHaveBeenCalledTimes(0)
  expect(experimentReloadRef.current).toHaveBeenCalledTimes(0)

  // Second Opening - We complete
  fireEvent.click(firstCompleteButton)

  await waitFor(() => screen.getByRole('button', { name: /Cancel/ }))
  const cancelButton2nd = screen.getByRole('button', { name: /Cancel/ })

  const allCompleteButtons = screen.getAllByRole('button', { name: /Complete/ })
  allCompleteButtons.forEach((button) => fireEvent.click(button))

  await waitForElementToBeRemoved(cancelButton2nd)

  expect(mockedExperimentsApi.changeStatus).toHaveBeenCalledTimes(1)
  expect(experimentReloadRef.current).toHaveBeenCalledTimes(1)
  expect(mockedExperimentsApi.changeStatus).toMatchInlineSnapshot(`
    [MockFunction] {
      "calls": Array [
        Array [
          1,
          "completed",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": Promise {},
        },
      ],
    }
  `)
})
