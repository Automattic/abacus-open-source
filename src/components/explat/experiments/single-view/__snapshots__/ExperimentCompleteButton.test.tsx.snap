// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`completes an experiment 1`] = `
<div
  aria-hidden="true"
>
  <span
    class=""
    title="Stop new user assignments. Current assignments remain active."
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-dangerButtonOutlined-6"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        2. Complete
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </span>
</div>
`;

exports[`renders as expected 1`] = `
<div>
  <span
    class=""
    title="This experiment is staging."
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-dangerButtonOutlined-2 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        2. Complete
      </span>
    </button>
  </span>
</div>
`;
